fedimint-client/src/transaction/builder.rs

```diff
@@ -2,28 +2,89 @@ use std::sync::Arc;
 
 use bitcoin30::key::KeyPair;
 use bitcoin30::secp256k1;
-use fedimint_core::core::{DynInput, DynOutput, IntoDynInstance, ModuleInstanceId};
+use fedimint_core::core::{DynInput, DynOutput, IInput, IntoDynInstance, ModuleInstanceId};
+use fedimint_core::task::{MaybeSend, MaybeSync};
 use fedimint_core::transaction::{Transaction, TransactionSignature};
 use fedimint_core::Amount;
 use itertools::multiunzip;
 use rand::{CryptoRng, Rng, RngCore};
 use secp256k1::Secp256k1;
 
 use crate::module::StateGenerator;
-use crate::sm::DynState;
+use crate::sm::{self, DynState};
+use crate::{
+    states_to_instanceless_dyn, InstancelessDynClientInput, InstancelessDynClientInputBundle,
+    InstancelessDynClientInputSM,
+};
 
 #[derive(Clone)]
-pub struct ClientInput<I = DynInput, S = DynState> {
+pub struct ClientInput<I = DynInput> {
     pub input: I,
     pub keys: Vec<KeyPair>,
     pub amount: Amount,
+}
+
+#[derive(Clone)]
+pub struct ClientInputSM<S = DynState> {
     pub state_machines: StateGenerator<S>,
 }
 
-impl<I, S> IntoDynInstance for ClientInput<I, S>
+/// A group of inputs and state machines responsible for driving their state
+///
+/// These must be kept together as a whole when including in a transaction.
+#[derive(Clone)]
+pub struct ClientInputBundle<I = DynInput, S = DynState> {
+    pub(crate) inputs: Vec<ClientInput<I>>,
+    pub(crate) sms: Vec<ClientInputSM<S>>,
+}
+
+impl<I, S> ClientInputBundle<I, S>
+where
+    I: IInput + MaybeSend + MaybeSync + 'static,
+    S: sm::IState + MaybeSend + MaybeSync + 'static,
+{
+    pub fn new(inputs: Vec<ClientInput<I>>, sms: Vec<ClientInputSM<S>>) -> Self {
+        Self { inputs, sms }
+    }
+
+    pub fn inputs(&self) -> &[ClientInput<I>] {
+        &self.inputs
+    }
+
+    pub fn sms(&self) -> &[ClientInputSM<S>] {
+        &self.sms
+    }
```

Comment:
Yes, we have some asserts etc. verifying number of notes, etc. Generally, since module code is deciding what goes into a bundle in a first place, being too paranoid about hiding it is not needed. These two getters give a RO-only access, so should be OK to prevent some silly mistakes.